apex/functions/sign-xpi/sign_xpi.py

```diff
@@ -0,0 +1,255 @@
+import base64
+import hashlib
+import logging
+import os.path
+import rfc822
+import sys
+import tempfile
+import zipfile
+
+import boto3
+import json
+import marshmallow.fields
+import rdflib
+import requests
+from requests_hawk import HawkAuth
+from signing_clients.apps import JarExtractor
+from six.moves.urllib.parse import urljoin
+
+CHUNK_SIZE = 512 * 1024
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+s3 = boto3.resource('s3')
+
+class SignXPIError(Exception):
+    """Abstract base class for errors in this lambda."""
+
+
+class ChecksumMatchError(SignXPIError):
+    def __init__(self, url, expected_checksum, actual_checksum):
+        message = "When fetching {}, expected checksum {} (got {})".format(
+            url, expected_checksum, actual_checksum)
+        super(ChecksumMatchError, self).__init__(message)
+        self.url = url
+        self.expected_checksum = expected_checksum
+        self.actual_checksum = actual_checksum
```

Comment:
No specific use case in mind, just wanting to make error information as structured as possible. But I can remove them if you want.